data-prepper-expression/src/test/java/org/opensearch/dataprepper/expression/util/LiteralMatcher.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.expression.util;
+
+import org.antlr.v4.runtime.tree.ParseTree;
+import org.hamcrest.Description;
+import org.hamcrest.DiagnosingMatcher;
+import org.hamcrest.Matcher;
+import org.opensearch.dataprepper.expression.antlr.DataPrepperExpressionParser;
+
+import static org.opensearch.dataprepper.expression.util.ContextMatcher.hasContext;
+import static org.opensearch.dataprepper.expression.util.TerminalNodeMatcher.isTerminalNode;
+
+public class LiteralMatcher extends SimpleExpressionMatcher {
+    private final Matcher<ParseTree> literalMatcher = hasContext(DataPrepperExpressionParser.LiteralContext.class, isTerminalNode());
+
+    //region valid rule order
+    private static final RuleClassOrderedList VALID_LITERAL_RULE_ORDER = new RuleClassOrderedList(
+            DataPrepperExpressionParser.ExpressionContext.class,
+            DataPrepperExpressionParser.ConditionalExpressionContext.class,
+            DataPrepperExpressionParser.EqualityOperatorExpressionContext.class,
+            DataPrepperExpressionParser.RegexOperatorExpressionContext.class,
+            DataPrepperExpressionParser.RelationalOperatorExpressionContext.class,
+            DataPrepperExpressionParser.SetOperatorExpressionContext.class,
+            DataPrepperExpressionParser.UnaryOperatorExpressionContext.class,
+            DataPrepperExpressionParser.PrimaryContext.class,
+            DataPrepperExpressionParser.LiteralContext.class
```

Comment:
I think some of the expressions here can be nested, like `ConditionalExpressionContext`, `EqualityOperatorExpressionContext`, `RelationalOperatorExpressionContext`, `SetOperatorExpressionContext`.  Can you help me understand how are we validating static order for literal?

Reply:
All of these classes can be nested. This matcher is a recursive check verifying the order of the nested classes are in a valid order and end with a terminal node. The use case is a shorter, easier to read matcher that can assert highly nested terminal nodes.

A tree like this should match as is valid:
```
ExpressionContext
  |
ConditionalExpressionContext
  |
EqualityOperatorExpressionContext
  |
RegexOperatorExpressionContext
  |
RelationalOperatorExpressionContext
  |
SetOperatorExpressionContext
  |
UnaryOperatorExpressionContext
  |
PrimaryContext
  |
LiteralContext
  |
TerminalNode
```

This tree will also match:
```
PrimaryContext
  |
LiteralContext
  |
TerminalNode
```

This tree will fail:
```
RegexOperatorExpressionContext
  |
RelationalOperatorExpressionContext
  |
SetOperatorExpressionContext
  |
UnaryOperatorExpressionContext
  |
PrimaryContext
  |
JsonPointer
```